modules/cas_cache/layer_cache_management.c

```diff
@@ -2042,30 +2069,45 @@ static int _cache_mngt_check_metadata(struct ocf_mngt_cache_config *cfg,
 	return result;
 }
 
-static int _cache_start_finalize(ocf_cache_t cache)
+static int _cache_start_finalize(ocf_cache_t cache,
+		struct kcas_start_cache *cmd)
 {
 	struct cache_priv *cache_priv = ocf_cache_get_priv(cache);
 	struct _cache_mngt_attach_context *ctx = cache_priv->attach_context;
 	int result;
 
 	_cache_mngt_log_cache_device_path(cache, ctx->device_cfg);
 
-	result = cas_cls_init(cache);
-	if (result) {
-		ctx->ocf_start_error = result;
-		return result;
+	if (cmd->init_cache != CACHE_INIT_BIND) {
+		result = cas_cls_init(cache);
+		if (result) {
+			ctx->ocf_start_error = result;
+			return result;
+		}
+		ctx->cls_inited = true;
 	}
-	ctx->cls_inited = true;
 
-	result = cache_mngt_initialize_core_objects(cache);
-	if (result) {
-		ctx->ocf_start_error = result;
-		return result;
+	switch(cmd->init_cache) {
+	case CACHE_INIT_LOAD:
+		result = cache_mngt_initialize_core_exported_objects(cache);
+		if (result) {
+			ctx->ocf_start_error = result;
+			return result;
+		}
+		ocf_core_visit(cache, _cache_mngt_core_device_loaded_visitor,
+				NULL, false);
+		break;
+	case CACHE_INIT_BIND:
+		result = cache_mngt_initialize_cache_exported_object(cache);
+		if (result) {
+			ctx->ocf_start_error = result;
+			return result;
+		}
+		break;
+	default:
```

Comment:
That was the intention. I had a choice between `switch` and `if`-`else if`, so I went for `switch` as I found it more elegant, but now I see it can be also confusing.

Reply:
Anyway, any other case in this situation actually one case, `CACHE_INIT_NEW`, so it may be easy to fix.